sampling/rate_limited/lib/opentelemetry/sampling/rate_limited/sampler.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Sampling
+    module RateLimited
+      class Sampler
+        def initialize(credits_per_second)
+          @credits_per_second = credits_per_second
+          @balance = 0
+          @last_tick = Time.now.to_i
+          @lock = Mutex.new
+        end
+
+        # @param [String] trace_id
+        # @param [OpenTelemetry::Context] parent_context
+        # @param [Enumerable<Link>] links
+        # @param [String] name
+        # @param [Symbol] kind
+        # @param [Hash<String, Object>] attributes
+        # @return [OpenTelemetry::SDK::Trace::Samplers::Result] The sampling result.
+        def should_sample?(trace_id:, parent_context:, links:, name:, kind:, attributes:)
+          if can_spend?
+            OpenTelemetry::SDK::Trace::Samplers::Result.new(
+              decision: OpenTelemetry::SDK::Trace::Samplers::Decision::RECORD_AND_SAMPLE,
+              tracestate: OpenTelemetry::Trace.current_span(parent_context).context.tracestate
+            )
+          else
+            OpenTelemetry::SDK::Trace::Samplers::Result.new(
+              decision: OpenTelemetry::SDK::Trace::Samplers::Decision::DROP,
+              tracestate: OpenTelemetry::Trace.current_span(parent_context).context.tracestate
+            )
+          end
+        end
+
+        private
+
+        # @return [Boolean]
+        def can_spend?
```

Comment:
Do we care about how this performs in the case where an app is configured with Puma, with Unicorn? Forking vs Pre-forking web servers will result in different spend rates. What's the expected outcome here?

Reply:
This is a good point that I haven't considered. I believe that the best option is to not make any assumptions about this at all. Configuring the rate accordingly could restore the parity between forking and pre-forking servers. Should we note this in the README?

Reply:
It's more that I think we want clarity on what the expectations are for the scope of the rate limiter.

Does the rate limit apply per process, per thread, or per fiber?

Reply:
Per process makes most sense, I believe. I don't see why different processes should have this knowledge of each other. It also makes the handling easier as we don't need to know the details of different servers